wordpress-mobile/WordPress-Android
# Add Tenor support

Summary
==========
This pull request proposes a solution that brings a complete integration for the Tenor API within the pre-existent GIF selection code inside the WordPress application, solving the [issue 11456](https://github.com/wordpress-mobile/WordPress-Android/issues/11456).

This is an aggregation branch for four different branches:

| Pull Request  | Link |
| ------------- | ------------- |
| Tenor API integrated with GIF Picker  | https://github.com/ThomazFB/WordPress-Android/pull/1  |
| GIF selection fully restored with Tenor API improvements  | https://github.com/ThomazFB/WordPress-Android/pull/2  |
| Replace UI and internal Giphy references  | https://github.com/ThomazFB/WordPress-Android/pull/3  |
| Add Tenor feature flag through selected build type  | https://github.com/ThomazFB/WordPress-Android/pull/4  |

Changelog
------------

### Tenor integration changes ###

* A `GifProvider` interface is now the only coupling between all code for the GIF picker and the Tenor implementation, avoiding any reference to the Tenor library and making it easy to replace the Provider as we're doing now.
* The `TenorProvider` implements the `GifProvider` interface and is injected within the `DataSource` via Dagger, this way only the `AppplicationModule` knows about the `TenorProvider`. Also, the `TenorProvider` is covered by unit tests in the `TenorProviderTest` class.
* The already implemented pagination feature and list actions are all kept and integrated to work smoothly with the `GifProvider`, this integration happens through the available `PositionalDataSource`, The Tenor implementation makes sure that positioning tracking is respected.
* The original integration with the `MediaBrowserActivity` was restored, making GIF selection available again.
* The original integration with the `EditPostActivity` was restored, making GIF selection available again when writing or editing posts.
* Analytics tracking restored.

### UI changes ###

* The Giphy attribution icon was removed and replaced by the Tenor attribution icon.
* Every UI string with `Giphy` information was replaced with `Tenor`. 

### Refactoring and improvements changes ###

* Test suite class added to `GiphyPickerDataSource` in order to better enforce the business logic when consuming any `GifProvider` implementation.
* The `giphy` packages inside `ui` and `viewmodel` were both renamed to only `gif`.
* All classes with the `Giphy` naming convention were renamed to only `Gif` to better fit the agnostic gif provider implementation strategy.
* All comments and documentation with references to `Giphy` or the Giphy SDK were readapted to better fit the `GifProvider` structure.

### Build configuration changes ###

* The `BuildConfig` now receive the declaration of the boolean `TENOR_AVAILABLE` field directly from the `build.gradle`.
* The button availability to select the Tenor GIF picker follows the `TENOR_AVAILABLE` value, if the value is false, the button doesn't get configured and should never appear on the screen.
* The `TENOR_AVAILABLE` is set as `true` for the Wasabi and Zalpha builds and `false` for the Vanilla build.
* Removed the Giphy SDK  reference at maven repository in `build.gradle`.

Testing
==========
Before doing any tests, make sure that to create a Tenor API key and paste it associated with the `wp.tenor.api_key` as the value inside your `gradle.properties`

GIF picking from Media Browser
------------

![screencapture-1585095917110 2020-03-24 23_04_23](https://user-images.githubusercontent.com/5920403/77494177-d3876d00-6e23-11ea-9928-16cf37e264e3.gif)

### How to test ###

1. Go to the Media Browser
2. Click on the Add button at the top right. Select the `Choose from Tenor` option
3. Type in the search bar to query GIFs
4. Once inside the Gif Picker, make sure that `selection`, `preview` and `add` are all working as expected

GIF picking from the Editor
------------

![screencapture-1585276984938 2020-03-26 23_46_37](https://user-images.githubusercontent.com/5920403/77717245-b89d3000-6fbe-11ea-8b8b-1b91699e1276.gif)

### How to test ###

1. Click to create a new post
2. Click on the overflow button at the top right end of the screen
3. select the `Switch to classic editor` 
4. Click on the plus (+) button at the bottom left end of the screen
5. Select the `Choose from Tenor` option
6. Type in the search bar to query GIFs
7. Once inside the Gif Picker, make sure that `selection`, `preview` and `add` are all working as expected

Tenor GIF picker feature flag - Post Editor
------------
| Wasabi and Zalpha  | Vanilla |
| ------------- | ------------- |
| ![screenshot-1585788446924](https://user-images.githubusercontent.com/5920403/78199765-fd174880-7462-11ea-9f4d-d8e95d6b0afa.jpg)  | ![screenshot-1585788432154](https://user-images.githubusercontent.com/5920403/78199766-fdafdf00-7462-11ea-8dd8-b1b00994d0df.jpg)  |

### How to test with Wasabi and Zalpha build ###
1. Click to create a new post
2. Click on the overflow button at the top right end of the screen
3. select the `Switch to classic editor` 
4. Click on the plus (+) button at the bottom left end of the screen
5. Verify that the `Choose from Tenor` option is available
6. Type in the search bar to query GIFs
7. Once inside the Gif Picker, make sure that `selection`, `preview` and `add` are all working as expected

### How to test with Vanilla build ###
1. Click to create a new post
2. Click on the overflow button at the top right end of the screen
3. select the `Switch to classic editor` 
4. Click on the plus (+) button at the bottom left end of the screen
5. Verify that the `Choose from Tenor` option isn't there

Tenor GIF picker feature flag - Media Browser
------------

| Wasabi and Zalpha  | Vanilla |
| ------------- | ------------- |
| ![screenshot-1585788487864](https://user-images.githubusercontent.com/5920403/78199760-f983c180-7462-11ea-8ddf-aab5daf67ef5.jpg)  | ![screenshot-1585788395658](https://user-images.githubusercontent.com/5920403/78199769-fe487580-7462-11ea-9c50-6f8ee8bfbfa7.jpg) |

### How to test with Wasabi and Zalpha build ###
1. Go to the Media Browser
2. Click on the Add button at the top right
3. Verify that the `Choose from Tenor` option is available
4. Type in the search bar to query GIFs
5. Once inside the Gif Picker, make sure that `selection`, `preview` and `add` are all working as expected

### How to test with Vanilla build ###
1. Go to the Media Browser
2. Click on the Add button at the top right.
3. Verify that the `Choose from Tenor` option isn't there

PR submission checklist
==========
